CHANGELOG.md

```diff
@@ -4,11 +4,18 @@
 
 * `geoipupdate` now supports retrying on more types of errors
   such as HTTP2 INTERNAL_ERROR.
-* `HTTPReader` no longer retries on HTTP errors and therefore
-  `retryFor` was removed from `NewHTTPReader`.
 * Now `geoipupdate` doesn't requires the user to specify the config file
   even if all the other arguments are set via the environment variables.
   Reported by jsf84ksnf. GitHub #284.
+* `pkg/goipupdate/database` has been divided and rewritten into multiple packages,
+  which change the behaviour of the tool:
+
+  * `pkg/goipupdate/database` which is responsible for handling api calls to maxmind
+    servers and includes two new api calls to the `metadata` and `download` endpoints.
+  * `pkg/goipupdate/writer` which is responsible for writing databases to various targets.
+    It only supports writing to disk out of the box.
+  * `pkg/goipupdate/lock` which is responsible for synchronizing concurrent access to the tool.
```

Comment:
What do you think about taking more things private? I was thinking the public package API could include only the minimum required. That would be, I think, retrieving metadata about edition ID(s) and downloading an edition. I don't think we need to be providing stuff like the config, locking files, writing files, etc publicly. That kind of stuff could be internal to geoipupdate. Basically anything that could be private, I was hoping we could make private.

Reply:
Which part do you mean when you say the "maxmind api"?

The main benefit is that it would make maintenance simpler. From experience working on it, it has become more complicated and error prone once we began sharing much of the internals.

If someone wants to write to places other than disk, wouldn't they still be able to do that given we'd be working with an io.Reader/io.Writer now? Same with other custom things. I'm not sure I would expect an HTTP API client to include that kind of stuff. We've basically provided the building blocks of re-creating `geoipupdate`, whereas I think it would be reasonable for us to only provide the HTTP API client part and let users figure out the rest if they need something custom.

Reply:
> I am sorry I don't fully get the point. We'd still have to maintain practically the same components/code whether they're made public or not.

Differently though. We are constrained by backwards compatibility and the existing interface. And it has made the code significantly more complicated due to the abstractions. What geoipupdate does is pretty simple but the code has become painful to work with IMO. And that is largely due to the constraints imposed by providing a public API together with its poor design (not talking about yours - I haven't looked at it in detail yet).

> nicer if everything is exposed so people requiring minor customization won't have to recreate geoipupdate from scratch

That's true, but at this point my preference would be to step back from that given the drag on development and the quality of the public interface. I don't think a lot of people are depending on it, but they could continue to use older versions for now too and pull in what they need from those versions if necessary. That kind of stuff could go into a separate package/project too.

Maybe we could see if anyone else has any thoughts on the direction here.

Reply:
I haven't read through all the threads, but I am generally in favor of making as much internal as possible. As Will mentioned, this code has become pretty hard to maintain as we are quite restricted by providing a stable API for library users. Many things that would have been trivial changes turn out to be complicated PRs or require unnecessary major version releases, which confuse the primary CLI users.